### DISCUSSION TITLE
Seeking Feedback on Test Data in Playwright E2E Tests

### DISCUSSION THREAD

**Discussion by imanish003** [woocommerce/woocommerce#44957] [WooCommerce, Blocks, and Site Editing]
I've been recently working on writing end-to-end (E2E) tests for our Product Collection block. As part of this process, I have been thinking about our approach to managing test data, and I wanted to share my thoughts and gather your feedback.

I have added around 8 playwright tests that depend on the default data inserted by the [test-env-setup.sh](https://github.com/woocommerce/woocommerce-blocks/blob/d34b7e078cc12da21055e3b9b808bcfc8e46c6e8/tests/e2e-pw/bin/test-env-setup.sh) file. Following code inserts the sample products data:

```
wp-env run tests-cli "wp import wp-content/plugins/woocommerce/sample-data/sample_products.xml --authors=skip"

```

## Relying on Default Data - Is it the Best Approach?

Let's look at a sample test to illustrate my point:

```tsx
test( 'Products can be filtered based on product attributes like color, size etc.', async ( {
	pageObject,
} ) => {
	await pageObject.setProductAttribute( 'Color', 'Green' );
	await expect( pageObject.products ).toHaveCount( 3 );

	await pageObject.setProductAttribute( 'Size', 'Large' );
	await expect( pageObject.products ).toHaveCount( 1 );

	await pageObject.publishAndGoToFrontend();
	await expect( pageObject.products ).toHaveCount( 1 );
} );
```

This test operates under two main assumptions:

- The data inserted by the **[test-env-setup.sh](https://github.com/woocommerce/woocommerce-blocks/blob/d34b7e078cc12da21055e3b9b808bcfc8e46c6e8/tests/e2e-pw/bin/test-env-setup.sh)** script won't change.
- No data will be added by other tests or test files to avoid potential side effects.

However, these assumptions might not always hold true, making the tests less robust. Future changes in the sample data might require updates in our tests.

## **An Alternative: The Fixture Data Approach**

In our Puppeteer tests, we utilized 'fixture' data, inserted before running E2E tests via the **[setup.js](https://github.com/woocommerce/woocommerce-blocks/blob/d34b7e078cc12da21055e3b9b808bcfc8e46c6e8/tests/e2e/config/setup.js)** script, and removed afterward with the **[teardown.js](https://github.com/woocommerce/woocommerce-blocks/blob/d34b7e078cc12da21055e3b9b808bcfc8e46c6e8/tests/e2e/config/teardown.js)** script.

### **The Power of Fixture Data in Tests**

By storing our data in the **`fixture-data.js`** file, we can create utility functions like this:

```tsx
/**
 * Internal dependencies
 */
import { Products } from '../e2e/fixtures/fixture-data';

/**
 * Get products data by key from fixtures.
 */
export const getFixtureProductsData = ( key = '' ) => {
	if ( ! key ) {
		return Products();
	}
	return Products()
		.map( ( product ) => product[ key ] )
		.filter( Boolean );
};

```

This function can be imported and used like so:

```tsx
import {
	getFixtureProductsData,
} from '@woocommerce/blocks-test-utils';

const outOfStockCount = getFixtureProductsData( 'stock_status' ).filter(
			( status: string ) => status === 'outofstock').length;

```

This allows us to calculate, for example, the number of out-of-stock products, ensuring our tests will always pass, even if the fixture data changes.

```tsx
expect( count ).toHaveLength( outOfStockCount );

```

### **Open Questions**

- Should I go ahead and add fixture data for Playwright?
    - Adopting a fixture-based approach, as used in our old puppeteer tests, can make our tests more robust. By leveraging utility functions to work with fixture data, we ensure that our tests remain accurate even if the fixture data changes.
- While the use of **`fixture-data.js`** significantly improves the robustness of our tests, there may be cases where we only need data for a specific test. In such scenarios, it might be beneficial to provide utility methods that allow us to insert data specifically for that test and remove it once the test passes. This approach ensures that the test remains self-contained and does not rely on global or shared data. Considering this approach can help us achieve more granular control over test data and reduce potential side effects. What are your thoughts?

I look forward to hearing your thoughts and opinions on this. Your feedback and suggestions are highly appreciated! 🙏

**Comment by Aljullu** [comment]
Thanks for opening this discussion, @imanish003! What you propose makes sense to me, but at the same time, I'm not sure if we avoided the fixture approach in the Playwright tests intentionally for some reason. Pinging the folks who worked on Playwright tests initially: @gigitux @nielslange @opr @wavvves does any of you know?

**Reply by gigitux** [reply]
> Thanks for opening this discussion, @imanish003! What you propose makes sense to me, but at the same time, I'm not sure if we avoided the fixture approach in the Playwright tests intentionally for some reason

Not really. We just picked the fastest way to set up the store and start to work with Playwright.

**Comment by kmanijak** [comment]
> Should I go ahead and add fixture data for Playwright?
> Adopting a fixture-based approach, as used in our old puppeteer tests, can make our tests more robust. By leveraging utility functions to work with fixture data, we ensure that our tests remain accurate even if the fixture data changes.

I totally agree with this. Tests should give us deterministic results no matter if they are run alone or in a suite and in which order. What you wrote means the tests depend on each other and a slight change in one test or changing the order may cause another test to break. Or even worse, result in false positives.

> While the use of fixture-data.js significantly improves the robustness of our tests, there may be cases where we only need data for a specific test. In such scenarios, it might be beneficial to provide utility methods that allow us to insert data specifically for that test and remove it once the test passes. This approach ensures that the test remains self-contained and does not rely on global or shared data. Considering this approach can help us achieve more granular control over test data and reduce potential side effects. What are your thoughts?

While I agree in principle, I think it could be implemented with the first use case, not necessarily upfront.

**Reply by imanish003** [reply]
Thank you for your insights and for supporting the idea of adopting a fixture-based approach 🤝

Regarding the implementation of utility methods for inserting and removing data specific to each test, I appreciate your suggestion of waiting for a concrete use-case before doing so. It's indeed a more practical approach to address this requirement when it arises. This way, we can build the utility function based on actual needs and avoid over-engineering. Thanks again for your valuable feedback! 🙌🏻

**Reply by gigitux** [reply]
>  totally agree with this. Tests should give us deterministic results no matter if they are run alone or in a suite and in which order. What you wrote means the tests depend on each other and a slight change in one test or changing the order may cause another test to break. Or even worse, result in false positives.

Not really, with the current implementation of the infrastructure, if your E2E test does a side effect, it should be marked as adding the `.side_effects` suffix to avoid the test will run in parallel with other tests. Also, we should pay attention to cleaning up any side effects on the `afterAll` hook. (https://github.com/woocommerce/woocommerce-blocks/blob/34fb8c0cef76baab1a2b21cfdd033e86607e2a09/docs/contributors/e2e-guidelines.md/#L37-L44 and https://github.com/woocommerce/woocommerce-blocks/blob/34fb8c0cef76baab1a2b21cfdd033e86607e2a09/docs/contributors/e2e-guidelines.md/#L29-L30) 

All the E2E tests that don't have the suffix `.side_effects` will run in parallel. Check [this issue](https://github.com/woocommerce/woocommerce-blocks/issues/9632) for more details.

> Regarding the implementation of utility methods for inserting and removing data specific to each test, I appreciate your suggestion of waiting for a concrete use-case before doing so.

Yeah, agree with @imanish003. What should be the benefit? 🤔

**Reply by imanish003** [reply]
> Yeah, agree with @imanish003. What should be the benefit? 🤔

Our initial approach will be to rely on Fixture data to cover all tests. However, there may be instances in the future where the Fixture data might not suffice, leading to a necessity for test-specific data. Currently, it's uncertain if such scenarios will occur, so it seems reasonable to wait until a concrete need surfaces.

One potential case where we might need test-specific data is when we want to modify the test data. Given that alterations to the test data could introduce side effects, isolating this data to specific tests might be a more prudent approach.

Furthermore, I've been considering an alternative strategy where we wouldn't rely on fixture data at all. Instead, we could insert the necessary test data for each test file and remove it upon completion of all tests. This method would ensure that all tests are isolated and self-contained, potentially providing a more robust testing environment.

What do you think?

**Reply by gigitux** [reply]
I don't think that we should add/remove products based on the test. If fixture data might not suffice, we can easily fix them and update all the utils functions.

**Reply by dinhtungdu** [reply]
> there may be cases where we only need data for a specific test. In such scenarios, it might be beneficial to provide utility methods that allow us to insert data specifically for that test and remove it once the test passes. This approach ensures that the test remains self-contained and does not rely on global or shared data.

I think we can only say it reduces the side effect. The test isn't totally self-contained. For example, if we create a new product, those products will still affect other tests running in parallel at that time.

**Comment by gigitux** [comment]
Hi @imanish003, 
Thanks for this idea. Let me make sure I align with your idea. Are you proposing to write utils function like `outOfStockCount` based on the products we add to the database? If this is the idea, I agree with you. And I agree that using javascript to add the products to the database allows us to work with those utils functions easily. My only concern is the speed of the setup process. I was working on this PR to improve the speed for the set-up of the store: https://github.com/woocommerce/woocommerce-blocks/pull/9813.

In general, I agree that we should improve the E2E setup script and we should follow your idea.

**Reply by imanish003** [reply]
> Are you proposing to write utils function like outOfStockCount based on the products we add to the database?

You've got it right. The proposal is indeed to have a file similar to [fixture-data.js](https://github.com/woocommerce/woocommerce-blocks/blob/c196069795dd46469e655baee57a6cf221bd1bca/tests/e2e/fixtures/fixture-data.js). This file will contain all the necessary test data. We would then add this data to our database using the `global-setup.ts` file and remove it using the `global-teardown.ts` file located in our `tests/e2e-pw` directory.

With this setup, `fixture-data.js` would be available for use within our test files. As I explained in the original post, we could calculate something like `outOfStockCount` using data from `fixture-data.js`.

I hope that clarifies the process. Let me know if you have any further questions! 🙂

> My only concern is the speed of the setup process. I was working on this PR to improve the speed for the set-up of the store: https://github.com/woocommerce/woocommerce-blocks/pull/9813.

I took a quick look at your PR, and it seems that you will be generating SQL queries and using those queries directly to insert data into the database instead of using a command like `wp-env run tests-cli ...`. Is that correct?

**Reply by gigitux** [reply]
Yeah, I agree with you! It makes sense to open an issue and assign it to this epic https://github.com/woocommerce/woocommerce-blocks/issues/9001.

**Reply by imanish003** [reply]
Sounds good! I will open an issue and assign it to the Epic.

How do you feel about the following approach?


Should we remove [this line](https://github.com/woocommerce/woocommerce-blocks/blob/c8014236531f8dc8fdc7d8cc84677ff52f729ab8/tests/e2e-pw/bin/test-env-setup.sh#L40) from `test-env-setup.sh`?
```
wp-env run tests-cli "wp import wp-content/plugins/woocommerce/sample-data/sample_products.xml --authors=skip"
```
This line relies on the `sample_products.xml `file from the `WooCommerce plugin`.

Instead, we could move all of this sample data to our new `fixture-data.js`. That way, `fixture-data.js` would become the single source of truth for our product data. Alternatively, do you think it's okay to continue using `wp-content/plugins/woocommerce/sample-data/sample_products.xml`?

**Reply by gigitux** [reply]
> Instead, we could move all of this sample data to our new fixture-data.js. That way, fixture-data.js would become the single source of truth for our product data.

I would proceed in this way:
- create the fixture-data.js.
- create a script that setup the env via the fixture-data.js and in the end export a woocommerce.sql file.
- E2E test env will import the woocommerce.sql file.


During my tests, importing products/configuration via .sql is very fast. Every time that the fixtures are updated, we have to generate a new woocommerce.sql file. Does it make sense?

**Reply by imanish003** [reply]
> create a script that setup the env via the fixture-data.js and in the end export a woocommerce.sql file.

May I know what you mean by `setup the env`? Can you provide me an example? AFAIK We will be inserting all the data using woocommerce.sql file which will be generated using `fixture-data.js`. Is there anything else that `fixture-data.js` file is supposed to do? 

> During my tests, importing products/configuration via .sql is very fast. Every time that the fixtures are updated, we have to generate a new woocommerce.sql file.

Makes sense to me.

**Reply by gigitux** [reply]
Mmm, it would not be very easy to generate a SQL file from fixture-data. In my idea, we should continue to insert the products in a WordPress website and, after that, export the database and get the .sql file. I mean this with: `setup the env`. Obviously, we will do this operation only when something in the fixture is changed.

**Reply by imanish003** [reply]
Thank you for the explanation, it makes sense now. Let me summarize my comprehension to ensure we're on the same page:

1. The test data insertion into the WordPress website will be facilitated via two files:
   - `test-env-setup.sh`: We'll continue using this file as our primary source for inserting existing test data into the WordPress site.
   - `fixture-data.js`: This new file will be employed for inserting additional fixture data. For Puppeteer tests, we're currently using `WooCommerceRestApi` to populate data into the WordPress site as evidenced [here](https://github.com/woocommerce/woocommerce-blocks/blob/de142fc4587561f9d4e401cdc13db6b7795e14cc/tests/e2e/fixtures/fixture-loaders.js#L23-L34). I infer that we plan to replicate this strategy for inserting fixture data in the Playwright E2E tests. Can you confirm this?

After the data has been completely set up on the WordPress site, we plan to export the database to an SQL file (let's call it `<file>.sql`) using a command such as `wp db export <file>.sql`.

2. Subsequently, we'll use `<file>.sql` to import the data into our test environment.

The generation of the `<file>.sql` would only be necessary when there's an update in the fixture data, ensuring our tests always run against the most recent dataset.

Could you please confirm whether my understanding aligns with your proposal?

**Reply by gigitux** [reply]
> I infer that we plan to replicate this strategy for inserting fixture data in the Playwright E2E tests. Can you confirm this?

I don't know, maybe there is a better way. I think that it makes sense to explore different approaches!

Yes, exactly! The only thing that I'm not sure about is point 1. I think that we should use only the `fixture-data.js` when it is ready to be used, no?

**Reply by imanish003** [reply]
> I don't know, maybe there is a better way. I think that it makes sense to explore different approaches!

Sure, That makes sense to me as well.

> The only thing that I'm not sure about is point 1. I think that we should use only the fixture-data.js when it is ready to be used, no?

Understood, let me clarify to ensure I've captured your proposition accurately:

- We should phase out the use of `test-env-setup.sh` entirely for Playwright tests, in favor of exclusively utilizing `fixture-data.js`.
    - The existing functionalities of `test-env-setup.sh` such as:
        - Emptying the site
        - Adding attributes
        - Import sample products and regenerate product lookup tables
        - Create pages and posts
        - Set up payment methods, and so on,
        should be transferred to `fixture-data.js`.
    - Post migration, `fixture-data.js` would not only perform the responsibilities of the now obsolete `test-env-setup.sh`, but also include the additional data necessary for the Playwright E2E tests.

Is my interpretation consistent with your proposal?

**Reply by gigitux** [reply]
Yes 👍

**Reply by imanish003** [reply]
Created issue #10189 & assigned it to epic #9001.
